File: Sources/Minutes/Providers/ShareClient.swift
import Foundation
import MinutesCore
import Security

struct ShareClient {
    static let baseURL = URL(string: "https://redrule.n3el.dev")!

    private struct Payload: Encodable {
        struct Segment: Encodable {
            let speaker: String
            let time: String
            let text: String
        }
        let note: MeetingNote
        let transcript: [Segment]?
    }

    func publish(_ share: MeetingShare, note: MeetingNote, transcript: [TranscriptSegment]) async throws {
        let payload = Payload(note: note, transcript: share.includesTranscript ? transcript.map {
            Payload.Segment(speaker: $0.speakerLabel, time: TranscriptMerger.timestamp($0.start), text: $0.text)
        } : nil)
        let data = try JSONEncoder().encode(payload)
        guard data.count <= 2_000_000 else { throw ShareError.message("This meeting is too large to share.") }
        try await request(share: share, method: "PUT", body: data)
    }

    func revoke(_ share: MeetingShare) async throws {
        try await request(share: share, method: "DELETE", body: nil)
    }

    private func request(share: MeetingShare, method: String, body: Data?) async throws {
        // Always use our configured host; stored links cannot redirect upload credentials.
        var components = URLComponents(url: Self.baseURL.appendingPathComponent("api/share"), resolvingAgainstBaseURL: false)!
        components.queryItems = [URLQueryItem(name: "id", value: share.id)]
        var request = URLRequest(url: components.url!)
        request.httpMethod = method
        request.timeoutInterval = 45
        request.setValue("Bearer \(try credential())", forHTTPHeaderField: "Authorization")
        request.setValue("application/json", forHTTPHeaderField: "Content-Type")
        request.httpBody = body
        let (data, response) = try await URLSession.shared.data(for: request)
        guard let response = response as? HTTPURLResponse, (200..<300).contains(response.statusCode) else {
            let error = try? JSONDecoder().decode(ServiceError.self, from: data)
            throw ShareError.message(error?.error ?? "Sharing could not complete. Check your connection and try again.")
        }
    }

    private struct ServiceError: Decodable { let error: String }

    private func credential() throws -> String {
        for service in ["Redrule", "Minutes"] {
            let query: [String: Any] = [kSecClass as String: kSecClassGenericPassword,
                                       kSecAttrService as String: service, kSecAttrAccount as String: "sharing",
                                       kSecReturnData as String: true, kSecMatchLimit as String: kSecMatchLimitOne]
            var result: CFTypeRef?
            let status = SecItemCopyMatching(query as CFDictionary, &result)
            if status == errSecItemNotFound { continue }
            guard status == errSecSuccess, let data = result as? Data, let key = String(data: data, encoding: .utf8) else {
                throw KeychainError.status(status)
            }
            return key
        }
        throw ShareError.message("Sharing is not configured on this Mac. Run the sharing setup script from the Redrule project.")
    }
}

enum ShareError: LocalizedError {
    case message(String)
    var errorDescription: String? { if case .message(let message) = self { return message }; return nil }
}

File: src-tauri/src/providers/share_client.rs
//! Publishes and revokes read-only links on the Redrule sharing service.
use std::time::Duration;

use serde::{Deserialize, Serialize};

use super::{NetResult, keychain};
use crate::core::models::{MeetingNote, MeetingShare, TranscriptSegment};
use crate::core::transcript::timestamp;
use crate::core::{Error, Result};

pub const BASE_URL: &str = "https://redrule.vercel.app";
const MAX_PAYLOAD: usize = 2_000_000;
const TIMEOUT: Duration = Duration::from_secs(45);

#[derive(Serialize)]
struct Payload<'a> {
    note: &'a MeetingNote,
    #[serde(skip_serializing_if = "Option::is_none")]
    transcript: Option<Vec<Segment>>,
}

#[derive(Serialize)]
struct Segment {
    speaker: String,
    time: String,
    text: String,
}

#[derive(Deserialize)]
struct ServiceError {
    error: String,
}

pub fn share_url(id: &str) -> String {
    format!("{BASE_URL}/s/{id}")
}

/// Share ids are 64 lowercase hex characters: unguessable, and safe in a URL.
pub fn new_share_id() -> String {
    format!("{}{}", uuid::Uuid::new_v4().simple(), uuid::Uuid::new_v4().simple())
}

pub async fn publish(http: &reqwest::Client, share: &MeetingShare, note: &MeetingNote, transcript: &[TranscriptSegment]) -> Result<()> {
    let transcript = share.includes_transcript.then(|| {
        transcript
            .iter()
            .map(|s| Segment { speaker: s.speaker_label(), time: timestamp(s.start), text: s.text.clone() })
            .collect()
    });
    let body = serde_json::to_vec(&Payload { note, transcript })?;
    if body.len() > MAX_PAYLOAD {
        return Err(Error::message("This meeting is too large to share."));
    }
    request(http, share, reqwest::Method::PUT, Some(body)).await
}

pub async fn revoke(http: &reqwest::Client, share: &MeetingShare) -> Result<()> {
    request(http, share, reqwest::Method::DELETE, None).await
}

async fn request(http: &reqwest::Client, share: &MeetingShare, method: reqwest::Method, body: Option<Vec<u8>>) -> Result<()> {
    // Always use our configured host; stored links cannot redirect upload credentials.
    let mut request = http
        .request(method, format!("{BASE_URL}/api/share"))
        .query(&[("id", &share.id)])
        .timeout(TIMEOUT)
        .bearer_auth(keychain::sharing_key()?)
        .header("Content-Type", "application/json");
    if let Some(body) = body {
        request = request.body(body);
    }
    let response = request.send().await.net()?;
    if response.status().is_success() {
        return Ok(());
    }
    let message = response.json::<ServiceError>().await.map(|e| e.error);
    Err(Error::message(message.unwrap_or_else(|_| "Sharing could not complete. Check your connection and try again.".into())))
}
